java/arcs/core/host/ServiceLoaderHostRegistry.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.host
+
+import arcs.core.sdk.Particle
+import java.util.ServiceLoader
+import kotlin.sequences.asSequence
+
+/**
+ * A HostRegistry that discovers the available [ArcHost]s available on this platform by using
+ * Java ServiceLoader and SPI.
+ */
+class ServiceLoaderHostRegistry() : HostRegistry {
+    val hosts: MutableList<ArcHost> = mutableListOf()
+
+    init {
+        loadAndRegisterHostsAndParticles().forEach { host -> registerHost(host) }
+    }
+
+    companion object {
+        fun instance() = ServiceLoaderHostRegistry()
+    }
+
+    override fun availableArcHosts(): List<ArcHost> {
+        return hosts
+    }
+
+    private fun loadAndRegisterHostsAndParticles(): List<ArcHost> {
+        // Load all @AutoService(Particle.class) types
+        val allParticles = ServiceLoader.load(Particle::class.java).iterator().asSequence()
+            .map { p -> p.javaClass }.toList()
```

Comment:
nit: the body of the `map` lambda could just be `{ it.javaClass }`
```suggestion
            .map { it.javaClass }.toList()
```